### ISSUE TITLE
Edge Cases for filename parser, detects a book as both Volume and Chapter

### ISSUE THREAD

**Issue by ocgineer** [Kareadita/Kavita#1534]
Documenting edge cases for the filename parser, as I know @tjarls is making improvements regarding the parser, maybe there are some ideas regarding this and/or he can include in his testing so there is no regression.

# Correctly shows as volumes, no chapters so these are parsed _correctly_.
```
2.5 Dimensional Seduction v01 (2022) (Digital) (1r0n).cbz
7 Billion Needles v01 (2010) (digital) (XRA9).cbr
7thGARDEN v01 (2016) (Digital) (danke).cbz
10 Dance v01 (2019) (Digital) (danke-Empire).cbz
12 Beast v01 (2015) (Digital) (danke-Empire).cbz
86--EIGHTY-SIX v01 (2020) (Digital) (LuCaZ).cbz
1122 - For a Happy Marriage v01 (2019) (Digital) (Shizu).cbz
Alice 19th v01 (2003) (Digital) (ripper).cbz
Anti-Magic Academy - The 35th Test Platoon v01 (2017) (Digital) (danke-Empire).cbz
Angels of Death Episode.0 v01 (2019) (Digital) (LuCaZ).cbz
Chobits - 20th Anniversary Edition v01 (2020) (Digital) (danke-Empire).cbz 
I Was Reincarnated as the 7th Prince so I Can Take My Time Perfecting My Magical Ability v01 (2021) (Digital) (danke-Empire).cbz
```
# Shows up as volume AND chapter (chapter is always the number in the series title).
```
ACCA - 13-Territory Inspection Department v01 (2017) (Digital) (Uasaha).cbz
ACCA - 13-Territory Inspection Department P.S. v01 (2020) (Digital) (LuCaZ).cbz
Brave 10 v01 (2013) (Digital) (Lovag-Empire).cbz
Black Panther and Sweet 16 v01 (2017) (Digital) (danke-Empire).cbz
Boys Over Flowers Season 2 v01 (2017) (Digital) (ripper).cbz
Chillin' in Another World with Level 2 Super Cheat Powers v01 (2021) (Digital) (danke-Empire + nao).cbz
Cyborg 009 v01 (2003) (Digital) (BlurPixel-Empire).cbr
Dementia 21 v01 (2018) (digital) (tunafan).cbr
Golgo 13 v01 - Supergun (2006) (Digital) (danke-Empire).cbz
GTO - 14 Days in Shonan v01 (2022) (Digital) (danke-Empire).cbz
Level 1 Demon Lord & One Room Hero v01 (2021) (Digital) (danke-Empire).cbz
My Unique Skill Makes Me OP Even at Level 1 v01 (2021) (Digital) (danke-Empire) (F).cbz
Otherworldly Munchkin - Let's Speedrun the Dungeon with Only 1 HP! v01 (2020) (Digital) (danke-Empire) (F).cbz
```
# Edge case of edge case (pretty sure due to the No.)
```
Kaiju No. 8 v01 (2021) (Digital) (1r0n) (f).cbz -- volumes show also up as chapter 8
Kaiju No. 8 c027 (2021) (Digital) (1r0n).cbz -- adding c resolves chapter detection
No. 5 v01 (2021) (Digital) (1r0n) (f2).cbz
No. 6 v1 (2015) (Digital) (ripper).cbz
```
# Ripper name c1fi7
Any filename with `(c1fi7)` will always show up as volume AND chapter 1, though the ripper is also between parentheses. And yes, there are currently quite some ripped manga with this rippers tag.

Other parsing issues #1336

**Comment by ocgineer** [comment]
Idealy would be something like this;
If a volume is found, and a loose number, but that number does not have a chapter marker (ch, chapter, c) and is before the [first] volume marker, assume it is part of the series title.

For the chapters it is already required to have a chapter marker if the series name contains a number, it will then show up correctly, even for Kaiju No. 8 chapters.

**Comment by ocgineer** [comment]
Another tricky edge case, having 'part x' in the sub-title of a volume.
```
Path of the Assassin v09 - Battle for Power Part 1 (2008) (Digital) (Lovag-Empire).cbz
Path of the Assassin v10 - Battle for Power Part 2 (2008) (Digital) (Lovag-Empire).cbz
```
Show up as volume 9 and 10 but also as chapter 1 and 2.
In this case maybe if 'part' is in front of the number, ignore as chapter. usually this is part of the title or sub-title. An example of it being in the title is;
```
Ascendance of a Bookworm - Part 01 v01 (2019) (Digital) (Ushi).cbz
...
Ascendance of a Bookworm - Part 02 v01 (2021) (Digital) (Ushi).cbz
```

**Comment by majora2007** [comment]
In this case the part doesn't make sense, but I'm webtoon I've seen parts used for chapter numbers. We'd need to investigate and prove it's not as common a case so we get the most benefit from file parsing (esp webtoon usually have even less metadata than manga)

**Comment by ocgineer** [comment]
So far for all the manga I have, the Path of the Assassin had Part x in the sub-title. the issue can be easily solved by just removing it from the filename. I just wanted to document that it is a possibility that it happens.

The only manga so far that has Part in the series title is Ascendence of a Bookworm, mimicking the Light Novel sub-titles.

I'll keep an eye out at my webtoons but so far I do not believe I have come across this, the officially translated rips so far have matched the naming scheme of ripped manga.

**Comment by ocgineer** [comment]
I did a quick test with ComicInfo, the first test with 
```
Kaiju No. 8 v01 (2021) (Digital) (1r0n) (f).cbz
<ComicInfo>
    <Title>Kaiju No. 8 volume 1</Title>
    <Series>Kaiju No. 8</Series>
    <Count>2</Count>
    <Volume>1</Volume>
</ComicInfo>
```
Resulted in that still also a chapter 8 shows up, in addition to the volume 1. Count was used to check if Kavita read the ComicInfo setting the series to completed.

```
A Bride's Story v02 (2011) (Digital) (c1fi7).cbz
<ComicInfo>
    <Title>A Bride's Story volume 2</Title>
    <Series>A Bride's Story</Series>
    <Count>2</Count>
    <Volume>2</Volume>
</ComicInfo>
```
Still showed up as chapter 1 due to the (c1fi7) in the filename, even if there is ComicInfo.xml. Removing (c1f17) from the filename and rescanning the series, removed the chapters as expected. This means that regardless of ComicInfo, the filename parser still tries to parse out chapters.

**Comment by majora2007** [comment]
It's not count, it's number. You used the wrong field.

**Comment by ocgineer** [comment]
I only used Count to have it shown up as "completed" in Kavita, to make sure it read the ComicInfo.xml. I double checked with the wiki `If you have at least one "Count" defined within any ComicInfo from the series, and it is not 0, then Kavita will assume the Series is Completed. Otherwise, it will be assumed Ongoing.`.

**Comment by majora2007** [comment]
I misread your post. You're talking about adding comicinfo without issue
number to validate the issue number from parser remains, which of course
since comicinfo overrides what filename parser finds.

**Comment by ocgineer** [comment]
Right, I thought that the filename parser was completely ignored if ComicInfo was present, testing thus using ComicInfo for [manga volumes] that has a number in the series name, would resolve the issue, which it does not then unfortunately.

**Comment by StereotypicalCat** [comment]
Im trying to do a few fixes to the parser - Id like to ask for some elaboration on these:
```
ReZERO -Starting Life in Another World- Chapter 3 - Truth of Zero v01 (2017) (Digital) (LuCaZ).cbz
```
For this, see this test case as well for ParseChaptersTest:
```
[InlineData("Accel World Chapter 001 Volume 002", "1")]
```
These test cases would be contradictory right? Does any of you (@majora2007 or @ocgineer) have an opinion on what we should do? I see two solutions:
* We can give up on this test case, stating it as ambigious as Chapter 3 should be parsed as a Chapter, and its too bad
* We can state that file names need to follow the pattern of "Volume -> Chapter", if a chapter should be recognized, fixing test case two, but not 1

The other case I need elaboration on is:
```
Higurashi When They Cry - Arc 9 - Dice Killing Arc (2014) (Digital) (LuCaZ).cbz
```
Should we accept "arc" as a volume marker? unlike the previous case, theres no volume marker here, so I am not sure what this should be parsed as? :)

Other than this, the changes from the newer commits from my latest PR https://github.com/Kareadita/Kavita/pull/4819 seems to help the parser loads in the other cases.

**Comment by majora2007** [comment]
Hi @StereotypicalCat, let me try to explain how these unique cases work. 

`ReZERO -Starting Life in Another World- Chapter 3 - Truth of Zero v01 (2017) (Digital) (LuCaZ).cbz`. The Chapter 3 here is part of the Series Name. There should be no Chapter capture in reality, however the Parser is just not that good.

`[InlineData("Accel World Chapter 001 Volume 002", "1")]` This is a valid case. Ch X Vol Y is sometimes used in the scene, we cannot break this use-case. 

`Higurashi When They Cry - Arc 9 - Dice Killing Arc (2014) (Digital) (LuCaZ).cbz` Arc is also not a volume, it's part of the Series Name in this case. 

I believe you should narrow the scope of your Parser changes to detect a case like `Chapter X {random words} Volume Y` and disregard the Chapter marker. `Chapter X Volume Y`, `Volume X Y` (chapter Y), `Series 01` (chapter 01) are all valid naming conventions. 

Hope this helps.

**Comment by StereotypicalCat** [comment]
> Hi [@StereotypicalCat](https://github.com/StereotypicalCat), let me try to explain how these unique cases work.
> 
> `ReZERO -Starting Life in Another World- Chapter 3 - Truth of Zero v01 (2017) (Digital) (LuCaZ).cbz`. The Chapter 3 here is part of the Series Name. There should be no Chapter capture in reality, however the Parser is just not that good.
> 
> `[InlineData("Accel World Chapter 001 Volume 002", "1")]` This is a valid case. Ch X Vol Y is sometimes used in the scene, we cannot break this use-case.
> 
> `Higurashi When They Cry - Arc 9 - Dice Killing Arc (2014) (Digital) (LuCaZ).cbz` Arc is also not a volume, it's part of the Series Name in this case.
> 
> I believe you should narrow the scope of your Parser changes to detect a case like `Chapter X {random words} Volume Y` and disregard the Chapter marker. `Chapter X Volume Y`, `Volume X Y` (chapter Y), `Series 01` (chapter 01) are all valid naming conventions.
> 
> Hope this helps.

It does help a lot, and I guess theres also an argument for strict vs loose parsing.

I don't currently feel confident in choosing a strategy/solution for  these two special cases (Re:Zero Chapter 1/2/3 and the Higurashi one), but the other edge cases in this issue should be fixed by the changes in #4819 - maybe someone in the future can fix the last couple of cases from this issue as mentioned above :)

**Comment by majora2007** [comment]
Correct, there are just some gaps that might not be solvable without ComicInfo. It's just one of the side effects of how I built Kavita and how I want files to be presented (aka no filenames in the UI).
